Adds client js test for event-target

andreban opened the pull request:
- Split tests between `app` and `client`.
- Updates .babelrc to support `default` and `test` environments.
- Update package.json
  - Adds 'mocha-client` target to test the client js.
  - Renames `mocha` to `mocha-app`.
  - `test` target invokes both `mocha-client` and `mocha-app`.

This will resolve #372

sebastianbenz commented on test/client/js/event-target.js:
```diff
@@ -0,0 +1,126 @@
+describe('js.event-target', () => {
+  describe('#addEventLister', () => {
+    it('Add a single listener to a type', () => {
+      const eventTarget = new EventTarget();
```
extract the object creation into a before block

sebastianbenz commented on test/client/js/event-target.js:
```diff
@@ -0,0 +1,126 @@
+  describe('#addEventLister', () => {
+    it('Add a single listener to a type', () => {
+      const eventTarget = new EventTarget();
+      const callback = () => {};
```
move callback declaration into before block

sebastianbenz commented on test/client/js/event-target.js:
```diff
@@ -0,0 +1,126 @@
+      const callback = () => {};
+      eventTarget.addEventListener('type', callback);
+      const eventListeners = eventTarget.getEventListeners('type');
+      assert.notEqual(eventListeners, null);
```
this assert Is not needed

sebastianbenz commented on test/client/js/event-target.js:
```diff
@@ -0,0 +1,126 @@
+      const eventListeners = eventTarget.getEventListeners('type');
+      assert.notEqual(eventListeners, null);
+      assert.equal(eventListeners.size, 1);
+      assert.equal(eventListeners.has(callback), true);
```
I wouldn't test whether an callback exists (or whether event listeners has size 1. These are implementation details. Testing implementation details make tests fragile. Instead focus tests on the behavior that is visible from the outside -> event dispatching. I

andreban commented:
Rewrote the tests.

sebastianbenz commented on test/client/js/event-target.js:
```diff
@@ -0,0 +1,53 @@
+  it('Fires the correct callback', done => {
+    eventTarget.addEventListener('event-a', callbackA);
+    eventTarget.dispatchEvent({type: 'event-a'});
+    setTimeout(() => {
```
I don't think we need a timeout here.

ithinkihaveacat replied:
Looking at https://mochajs.org/#asynchronous-code and also https://lostechies.com/derickbailey/2012/08/17/asynchronous-unit-tests-with-mocha-promises-and-winjs/ I think the right way to do this is to pass `done` to the callback, and arrange for it to be called.

The problem with removing `setTimeout()` completely is that `dispatchEvent()` might (for some reason) only process events after a timeout, in which case the `done()` has already happened. Mocha apparently imposes its own 2 sec timeout for `done()` to happen, so I think this test should look like:

```js
it("Fires the correct callback 2", done => {
  const expectedEvent = { type: "event-a" };
  eventTarget.addEventListener("event-a", actualEvent => {
    assert.deepEqual(actualEvent, expectedEvent);
    done();
  });
  eventTarget.dispatchEvent(expectedEvent);
});
```

(`before()` should probably also be changed to `beforeEach()` to get a fresh `eventTarget` each time, as I discovered when adding a "Fires the correct callback 2" test...)

sebastianbenz replied:
How do you check if an event hasn't been fired in this case?

Why not simply go with:

```
it("Fires the correct callback 2", () => {
  eventTarget.addEventListener('event-a', callbackA);
  eventTarget.dispatchEvent({type: 'event-a'});
  assert.equal(callbackA.callCount, 1);
});
```

Guarding against a future change in eventTarget is difficult as we don't know how it's going to change.

ithinkihaveacat replied:
@andreban pointed that out as well. `done()` with an argument is considered an error, so that could possibly be used (instead of making the spy smarter). Getting rid of that `setTimeout` is the most important thing.

andreban commented:
Removed the timeouts. Fixed using beforeEach instead of before. Keeping the spy, as it makes the test clearer. Also, avoids issues with things such as the callback being invoked twice, which would fail the test with callcount = 2, but would pass when calling done on the first callback.
